File: Sources/AdbShell/Services/ADBService.swift
import Foundation

enum ADBError: LocalizedError {
    case adbNotFound
    case commandFailed(String)

    var errorDescription: String? {
        switch self {
        case .adbNotFound:
            return L("error.adbNotFound")
        case .commandFailed(let message):
            return message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? L("error.commandFailed") : message
        }
    }
}

struct ProcessResult {
    let stdout: String
    let stderr: String
    let exitCode: Int32

    var combined: String {
        [stdout, stderr].filter { !$0.isEmpty }.joined(separator: "\n")
    }
}

/// Тонкая обёртка над CLI `adb`, вызывающая процесс и парсящая его вывод.
final class ADBService {
    let adbPath: String

    init() {
        self.adbPath = ADBService.locateADB()
    }

    static func locateADB() -> String {
        // 1) adb, вшитый в .app сборкой build_app.sh — так релиз не требует
        //    отдельной установки Android Platform Tools на машине пользователя.
        if let bundled = Bundle.main.resourceURL?.appendingPathComponent("adb").path,
           FileManager.default.isExecutableFile(atPath: bundled) {
            return bundled
        }
        // 2) системный adb — для `swift run` в разработке и на случай,
        //    если пользователь всё же поставил Platform Tools сам.
        let candidates = [
            "/opt/homebrew/bin/adb",
            "/usr/local/bin/adb",
            "/usr/bin/adb"
        ]
        for path in candidates where FileManager.default.isExecutableFile(atPath: path) {
            return path
        }
        return "adb"
    }

    var isAvailable: Bool {
        FileManager.default.isExecutableFile(atPath: adbPath) || adbPath == "adb"
    }

    // MARK: - Низкоуровневый запуск

    @discardableResult
    func run(_ args: [String], serial: String? = nil, timeout: TimeInterval? = nil) async throws -> ProcessResult {
        let adbPath = self.adbPath
        return try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async {
                var allArgs: [String] = []
                if let serial {
                    allArgs += ["-s", serial]
                }
                allArgs += args

                let process = Process()
                process.executableURL = URL(fileURLWithPath: adbPath.hasPrefix("/") ? adbPath : "/usr/bin/env")
                process.arguments = adbPath.hasPrefix("/") ? allArgs : (["adb"] + allArgs)

                let outPipe = Pipe()
                let errPipe = Pipe()
                process.standardOutput = outPipe
                process.standardError = errPipe

                do {
                    try process.run()
                } catch {
                    continuation.resume(throwing: ADBError.commandFailed(L("error.adbLaunchFailed", error.localizedDescription)))
                    return
                }

                process.waitUntilExit()

                let outData = outPipe.fileHandleForReading.readDataToEndOfFile()
                let errData = errPipe.fileHandleForReading.readDataToEndOfFile()
                let out = String(data: outData, encoding: .utf8) ?? ""
                let err = String(data: errData, encoding: .utf8) ?? ""

                continuation.resume(returning: ProcessResult(stdout: out, stderr: err, exitCode: process.terminationStatus))
            }
        }
    }

    // MARK: - Устройства

    func listDevices() async throws -> [Device] {
        let result = try await run(["devices", "-l"])
        return Self.parseDevices(from: result.stdout)
    }

    /// Парсинг вывода `adb devices -l` — вынесено в чистую функцию для юнит-тестов.
    static func parseDevices(from output: String) -> [Device] {
        var devices: [Device] = []
        let lines = output.split(separator: "\n").map(String.init)
        for line in lines {
            if line.hasPrefix("List of devices") { continue }
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty { continue }
            let parts = trimmed.split(separator: " ", omittingEmptySubsequences: true).map(String.init)
            guard parts.count >= 2 else { continue }
            let serial = parts[0]
            let stateRaw = parts[1]
            let state = Device.State(rawValue: stateRaw) ?? .unknown

            var model: String?
            var product: String?
            var transportId: String?
            for token in parts.dropFirst(2) {
                if token.hasPrefix("model:") { model = String(token.dropFirst("model:".count)) }
                if token.hasPrefix("product:") { product = String(token.dropFirst("product:".count)) }
                if token.hasPrefix("transport_id:") { transportId = String(token.dropFirst("transport_id:".count)) }
            }
            devices.append(Device(serial: serial, state: state, model: model, product: product, transportId: transportId))
        }
        return devices
    }

    func connect(host: String) async throws -> String {
        let result = try await run(["connect", host])
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
        return result.stdout.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func disconnect(serial: String) async throws {
        let result = try await run(["disconnect", serial])
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
    }

    /// Сопряжение по коду для Android 11+ wireless debugging
    /// (Настройки → Для разработчиков → Отладка по Wi-Fi → Сопряжение с
    /// устройством по коду — там показаны host:port для пейринга и код).
    /// После успешного пейринга нужно ещё `connect` на отдельный порт
    /// подключения, который Android покажет на том же экране.
    func pair(hostPort: String, code: String) async throws -> String {
        let result = try await run(["pair", hostPort, code])
        let combined = result.combined
        if result.exitCode != 0 || combined.localizedCaseInsensitiveContains("failed") {
            throw ADBError.commandFailed(combined.isEmpty ? L("error.pairFailed") : combined)
        }
        return combined
    }

    /// Устройства с включённой беспроводной отладкой (Android 11+), найденные
    /// в локальной сети через mDNS/Bonjour — без ручного ввода IP.
    func discoverMdnsDevices() async throws -> [MdnsDevice] {
        let result = try await run(["mdns", "services"])
        return MdnsParser.parse(result.stdout)
    }

    // MARK: - Приложения

    func listApps(serial: String) async throws -> [InstalledApp] {
        async let allResult = run(["shell", "pm", "list", "packages"], serial: serial)
        async let userResult = run(["shell", "pm", "list", "packages", "-3"], serial: serial)
        async let disabledResult = run(["shell", "pm", "list", "packages", "-d"], serial: serial)

        let all = try await allResult
        let user = try await userResult
        let disabled = try await disabledResult

        return Self.mergeApps(all: all.stdout, user: user.stdout, disabled: disabled.stdout)
    }

    /// Склеивает три вывода `pm list packages` в список приложений с флагами
    /// isSystem/isEnabled — вынесено в чистую функцию для юнит-тестов.
    static func mergeApps(all: String, user: String, disabled: String) -> [InstalledApp] {
        func packageSet(from output: String) -> Set<String> {
            Set(output.split(separator: "\n").compactMap { line -> String? in
                let s = line.trimmingCharacters(in: .whitespaces)
                guard s.hasPrefix("package:") else { return nil }
                return String(s.dropFirst("package:".count))
            })
        }

        let allSet = packageSet(from: all)
        let userSet = packageSet(from: user)
        let disabledSet = packageSet(from: disabled)

        return allSet.map { pkg in
            InstalledApp(packageName: pkg, isSystem: !userSet.contains(pkg), isEnabled: !disabledSet.contains(pkg))
        }.sorted { $0.packageName.lowercased() < $1.packageName.lowercased() }
    }

    func appDetail(serial: String, packageName: String) async throws -> AppDetail {
        let result = try await run(["shell", "dumpsys", "package", packageName], serial: serial)
        return DumpsysParser.parseAppDetail(packageName: packageName, output: result.stdout)
    }

    func install(serial: String, apkPath: String) async throws -> String {
        let result = try await run(["install", "-r", "-g", apkPath], serial: serial, timeout: 120)
        let combined = result.combined
        if result.exitCode != 0 || combined.contains("Failure") {
            throw ADBError.commandFailed(combined)
        }
        return combined
    }

    func uninstall(serial: String, packageName: String) async throws {
        let result = try await run(["uninstall", packageName], serial: serial)
        let combined = result.combined
        if result.exitCode != 0 || combined.contains("Failure") {
            throw ADBError.commandFailed(combined)
        }
    }

    /// Пути к APK установленного пакета (`pm path`) — может быть несколько для split APK.
    func apkPaths(serial: String, packageName: String) async throws -> [String] {
        let result = try await run(["shell", "pm", "path", packageName], serial: serial)
        let paths = result.stdout.split(separator: "\n").compactMap { line -> String? in
            let s = line.trimmingCharacters(in: .whitespaces)
            guard s.hasPrefix("package:") else { return nil }
            return String(s.dropFirst("package:".count))
        }
        guard !paths.isEmpty else {
            throw ADBError.commandFailed(result.combined.isEmpty ? L("error.apkPathNotFound") : result.combined)
        }
        return paths
    }

    // MARK: - Файлы устройства

    func listDirectory(serial: String, path: String) async throws -> [RemoteFile] {
        let result = try await run(["shell", "ls", "-la", path], serial: serial)
        if result.stdout.isEmpty && !result.stderr.isEmpty {
            throw ADBError.commandFailed(result.stderr)
        }
        return RemoteFileParser.parse(output: result.stdout, parentPath: path)
    }

    func push(serial: String, localPath: String, remotePath: String) async throws {
        let result = try await run(["push", localPath, remotePath], serial: serial, timeout: 300)
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
    }

    func pull(serial: String, remotePath: String, localPath: String) async throws {
        let result = try await run(["pull", remotePath, localPath], serial: serial, timeout: 300)
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
    }

    func makeDirectory(serial: String, path: String) async throws {
        let result = try await run(["shell", "mkdir", "-p", path], serial: serial)
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
    }

    func removeRemote(serial: String, path: String, recursive: Bool) async throws {
        var args = ["shell", "rm"]
        args.append(recursive ? "-rf" : "-f")
        args.append(path)
        let result = try await run(args, serial: serial)
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
    }

    func forceStop(serial: String, packageName: String) async throws {
        try await run(["shell", "am", "force-stop", packageName], serial: serial)
    }

    func clearData(serial: String, packageName: String) async throws {
        let result = try await run(["shell", "pm", "clear", packageName], serial: serial)
        if result.stdout.contains("Failed") {
            throw ADBError.commandFailed(result.combined)
        }
    }

    func setEnabled(serial: String, packageName: String, enabled: Bool) async throws {
        let subcommand = enabled ? "enable" : "disable-user"
        var args = ["shell", "pm", subcommand]
        if !enabled { args += ["--user", "0"] }
        args.append(packageName)
        let result = try await run(args, serial: serial)
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
    }

    func grantPermission(serial: String, packageName: String, permission: String) async throws {
        let result = try await run(["shell", "pm", "grant", packageName, permission], serial: serial)
        if result.exitCode != 0 || !result.stderr.isEmpty {
            throw ADBError.commandFailed(result.combined)
        }
    }

    func revokePermission(serial: String, packageName: String, permission: String) async throws {
        let result = try await run(["shell", "pm", "revoke", packageName, permission], serial: serial)
        if result.exitCode != 0 || !result.stderr.isEmpty {
            throw ADBError.commandFailed(result.combined)
        }
    }

    /// Короткая строка вида "Android 13 · SDK 33" для шапки устройства.
    func buildFingerprint(serial: String) async throws -> String {
        async let releaseResult = run(["shell", "getprop", "ro.build.version.release"], serial: serial)
        async let sdkResult = run(["shell", "getprop", "ro.build.version.sdk"], serial: serial)
        let release = try await releaseResult.stdout.trimmingCharacters(in: .whitespacesAndNewlines)
        let sdk = try await sdkResult.stdout.trimmingCharacters(in: .whitespacesAndNewlines)
        var parts: [String] = []
        if !release.isEmpty { parts.append("Android \(release)") }
        if !sdk.isEmpty { parts.append("SDK \(sdk)") }
        return parts.joined(separator: " · ")
    }

    // MARK: - Мониторинг

    /// Снимок CPU/памяти/батареи устройства для вкладки "Мониторинг".
    /// Три независимые shell-команды запускаются параллельно, чтобы не растягивать
    /// интервал опроса — каждая по отдельности быстрая, но последовательно давали бы заметный лаг.
    func deviceStats(serial: String) async throws -> DeviceStats {
        async let cpuResult = run(["shell", "dumpsys", "cpuinfo"], serial: serial)
        async let memResult = run(["shell", "cat", "/proc/meminfo"], serial: serial)
        async let batteryResult = run(["shell", "dumpsys", "battery"], serial: serial)
        return DeviceStatsParser.parse(
            cpuOutput: try await cpuResult.combined,
            memOutput: try await memResult.combined,
            batteryOutput: try await batteryResult.combined
        )
    }

    /// Суммарные RX/TX байты приложения с последнего сброса счётчиков netstats
    /// (обычно с загрузки устройства). Вызывающая сторона сама считает дельту
    /// между двумя опросами, чтобы получить скорость.
    func networkUsage(serial: String, uid: Int) async throws -> (rxBytes: Int64, txBytes: Int64) {
        let result = try await run(["shell", "dumpsys", "netstats", "detail"], serial: serial)
        return NetworkUsageParser.parse(output: result.combined, uid: uid)
    }

    func runningProcesses(serial: String) async throws -> [RunningProcess] {
        let result = try await run(["shell", "ps", "-A", "-o", "PID,PPID,USER,RSS,NAME"], serial: serial)
        return ProcessListParser.parse(result.combined)
    }

    /// Убивает процесс по PID. Без root работает только для процессов того же UID,
    /// что и adb shell — на остальных вернёт permission denied, это ожидаемо.
    func killProcess(serial: String, pid: Int) async throws {
        let result = try await run(["shell", "kill", "-9", String(pid)], serial: serial)
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
    }

    // MARK: - ANR / tombstones

    /// Список файлов в `/data/anr/` и `/data/tombstones/`. Без root оба каталога
    /// обычно недоступны (`Permission denied`) — в этом случае просто пропускаем
    /// соответствующую директорию, а не считаем это ошибкой всей операции.
    func crashTraces(serial: String) async throws -> [CrashTraceFile] {
        async let anrResult = run(["shell", "ls", "-1", "/data/anr/"], serial: serial)
        async let tombResult = run(["shell", "ls", "-1", "/data/tombstones/"], serial: serial)

        var files: [CrashTraceFile] = []
        if let anr = try? await anrResult, anr.exitCode == 0 {
            files += parseListing(anr.stdout, dir: "/data/anr/", kind: .anr)
        }
        if let tomb = try? await tombResult, tomb.exitCode == 0 {
            files += parseListing(tomb.stdout, dir: "/data/tombstones/", kind: .tombstone)
        }
        return files
    }

    private func parseListing(_ stdout: String, dir: String, kind: CrashTraceFile.Kind) -> [CrashTraceFile] {
        stdout.split(separator: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty && $0 != "." && $0 != ".." }
            .map { CrashTraceFile(path: dir + $0, name: $0, kind: kind) }
    }

    /// Хвост файла трейса — полные tombstone-файлы могут быть большими,
    /// показываем последние ~30000 байт, обычно там самое важное (стек, сигнал).
    func readCrashTrace(serial: String, path: String) async throws -> String {
        let result = try await run(["shell", "tail", "-c", "30000", path], serial: serial)
        if result.exitCode != 0 {
            throw ADBError.commandFailed(result.combined)
        }
        return result.stdout
    }

    // MARK: - Logcat

    func makeLogcatSession(serial: String) -> LogcatSession {
        LogcatSession(adbPath: adbPath, serial: serial)
    }

    // MARK: - Прочее

    func shell(serial: String, command: String) async throws -> String {
        let result = try await run(["shell", command], serial: serial)
        return result.combined
    }

    func reboot(serial: String) async throws {
        try await run(["reboot"], serial: serial)
    }

    func rebootToRecovery(serial: String) async throws {
        try await run(["reboot", "recovery"], serial: serial)
    }

    func rebootToBootloader(serial: String) async throws {
        try await run(["reboot", "bootloader"], serial: serial)
    }

    /// `adb root` перезапускает adbd с правами root на устройстве — работает
    /// только на userdebug/eng сборках, на user-сборках вернёт ошибку.
    @discardableResult
    func rootAdb(serial: String) async throws -> String {
        let result = try await run(["root"], serial: serial)
        return result.combined
    }

    /// `adb remount` — перемонтирует системные разделы в RW (после adb root).
    @discardableResult
    func remount(serial: String) async throws -> String {
        let result = try await run(["remount"], serial: serial)
        return result.combined
    }

    func screenshot(serial: String) async throws -> Data {
        let adbPath = self.adbPath
        return try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async {
                let process = Process()
                process.executableURL = URL(fileURLWithPath: adbPath.hasPrefix("/") ? adbPath : "/usr/bin/env")
                var args = adbPath.hasPrefix("/") ? [String]() : ["adb"]
                args += ["-s", serial, "exec-out", "screencap", "-p"]
                process.arguments = args
                let outPipe = Pipe()
                process.standardOutput = outPipe
                process.standardError = Pipe()
                do {
                    try process.run()
                } catch {
                    continuation.resume(throwing: ADBError.commandFailed(error.localizedDescription))
                    return
                }
                let data = outPipe.fileHandleForReading.readDataToEndOfFile()
                process.waitUntilExit()
                continuation.resume(returning: data)
            }
        }
    }
}

File: desktop/src/main/adb/AdbService.ts
// Порт Sources/AdbShell/Services/ADBService.swift — тонкая обёртка над CLI
// `adb`, вызывающая процесс и парсящая его вывод. Намеренно не импортирует
// 'electron' — должен оставаться тестируемым напрямую через `node --test`,
// без запуска в Electron-рантайме.

import { spawn } from 'node:child_process';
import * as fs from 'node:fs';
import * as path from 'node:path';

import { Device } from './types/Device';
import { parseDevices } from './parsers/DeviceParser';
import { ProcessResult, combinedOutput } from './types/ProcessResult';
import { DeviceProperty, parseGetprop } from './parsers/GetpropParser';
import { PortForwardRule } from './types/PortForwardRule';
import { parseForwardList, parseReverseList } from './parsers/PortForwardParser';
import { parseDeviceIP } from './parsers/IpRouteParser';
import { AppDetail, InstalledApp } from './types/AppInfo';
import { mergeApps } from './parsers/AppListParser';
import { parseAppDetail } from './parsers/DumpsysParser';
import { RemoteFile } from './types/RemoteFile';
import { DeviceStats } from './types/DeviceStats';
import { parseDeviceStats } from './parsers/DeviceStatsParser';
import { RunningProcess } from './types/RunningProcess';
import { parseProcessList } from './parsers/ProcessListParser';
import { parseRemoteFiles } from './parsers/RemoteFileParser';
import { MdnsDevice } from './types/MdnsDevice';
import { parseMdnsServices } from './parsers/MdnsParser';
import { normalizeConnectHost } from './parsers/ConnectHost';
import { DeviceSecurityInfo } from './types/DeviceSecurityInfo';
import { NetworkUsage, parseNetworkUsage } from './parsers/NetworkUsageParser';
import { AppUsageStat } from './types/AppUsageStat';
import { parseUsageStats } from './parsers/UsageStatsParser';
import { CrashTraceFile } from './types/CrashTraceFile';
import { parseCrashTraceListing } from './parsers/CrashTraceParser';

export class AdbCommandError extends Error {}

export class AdbService {
  readonly adbPath: string;

  constructor(adbPath?: string) {
    this.adbPath = adbPath ?? AdbService.locateAdb();
  }

  /** Аналог ADBService.locateADB() — вшитый в приложение бинарник (Фаза 6
   * упаковки положит его в resources), иначе полагаемся на PATH. */
  static locateAdb(): string {
    const exeName = process.platform === 'win32' ? 'adb.exe' : 'adb';
    const resourcesPath = (process as unknown as { resourcesPath?: string }).resourcesPath;
    if (resourcesPath) {
      const bundled = path.join(resourcesPath, exeName);
      if (fs.existsSync(bundled)) return bundled;
    }
    return exeName;
  }

  run(args: string[], options: { serial?: string; timeoutMs?: number } = {}): Promise<ProcessResult> {
    const allArgs = options.serial ? ['-s', options.serial, ...args] : args;
    return new Promise((resolve, reject) => {
      let child;
      try {
        child = spawn(this.adbPath, allArgs, { windowsHide: true });
      } catch (error) {
        reject(new AdbCommandError(`Couldn't launch adb: ${(error as Error).message}`));
        return;
      }

      let stdout = '';
      let stderr = '';
      let settled = false;
      const timer = options.timeoutMs
        ? setTimeout(() => {
            if (!settled) child.kill();
          }, options.timeoutMs)
        : undefined;

      child.stdout?.on('data', (chunk: Buffer) => (stdout += chunk.toString('utf8')));
      child.stderr?.on('data', (chunk: Buffer) => (stderr += chunk.toString('utf8')));
      child.on('error', (error) => {
        if (settled) return;
        settled = true;
        if (timer) clearTimeout(timer);
        reject(new AdbCommandError(`Couldn't launch adb: ${error.message}`));
      });
      child.on('close', (code) => {
        if (settled) return;
        settled = true;
        if (timer) clearTimeout(timer);
        resolve({ stdout, stderr, exitCode: code });
      });
    });
  }

  // MARK: Устройства

  async listDevices(): Promise<Device[]> {
    const result = await this.run(['devices', '-l']);
    return parseDevices(result.stdout);
  }

  async connect(host: string): Promise<string> {
    const result = await this.run(['connect', normalizeConnectHost(host)]);
    return combinedOutput(result).trim();
  }

  async disconnect(serial: string): Promise<void> {
    await this.run(['disconnect', serial]);
  }

  async pair(hostPort: string, code: string): Promise<string> {
    const result = await this.run(['pair', hostPort, code], { timeoutMs: 15000 });
    return combinedOutput(result).trim();
  }

  /** Аналог ADBService.discoverMdnsDevices() — устройства с включённой
   * беспроводной отладкой (Android 11+), рекламирующие себя по mDNS. */
  async discoverMdnsDevices(): Promise<MdnsDevice[]> {
    const result = await this.run(['mdns', 'services']);
    return parseMdnsServices(result.stdout);
  }

  async shell(serial: string, command: string): Promise<string> {
    const result = await this.run(['shell', command], { serial });
    return combinedOutput(result);
  }

  // MARK: Wi-Fi отладка

  async enableWirelessDebugging(serial: string, port = 5555): Promise<string> {
    const result = await this.run(['tcpip', String(port)], { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
    return combinedOutput(result);
  }

  async deviceIPAddress(serial: string): Promise<string | undefined> {
    const result = await this.run(['shell', 'ip', 'route'], { serial });
    return parseDeviceIP(result.stdout);
  }

  // MARK: Проброс портов

  async listForwards(serial: string): Promise<PortForwardRule[]> {
    const result = await this.run(['forward', '--list'], { serial });
    return parseForwardList(result.stdout, serial);
  }

  async addForward(serial: string, hostSpec: string, deviceSpec: string): Promise<void> {
    const result = await this.run(['forward', hostSpec, deviceSpec], { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  async removeForward(serial: string, hostSpec: string): Promise<void> {
    const result = await this.run(['forward', '--remove', hostSpec], { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  async listReverses(serial: string): Promise<PortForwardRule[]> {
    const result = await this.run(['reverse', '--list'], { serial });
    return parseReverseList(result.stdout, serial);
  }

  async addReverse(serial: string, deviceSpec: string, hostSpec: string): Promise<void> {
    const result = await this.run(['reverse', deviceSpec, hostSpec], { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  async removeReverse(serial: string, deviceSpec: string): Promise<void> {
    const result = await this.run(['reverse', '--remove', deviceSpec], { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  // MARK: Свойства устройства

  async allProperties(serial: string): Promise<DeviceProperty[]> {
    const result = await this.run(['shell', 'getprop'], { serial });
    return parseGetprop(result.stdout);
  }

  // MARK: Приложения

  async listApps(serial: string): Promise<InstalledApp[]> {
    const [all, user, disabled] = await Promise.all([
      this.run(['shell', 'pm', 'list', 'packages'], { serial }),
      this.run(['shell', 'pm', 'list', 'packages', '-3'], { serial }),
      this.run(['shell', 'pm', 'list', 'packages', '-d'], { serial }),
    ]);
    return mergeApps(all.stdout, user.stdout, disabled.stdout);
  }

  async appDetail(serial: string, packageName: string): Promise<AppDetail> {
    const result = await this.run(['shell', 'dumpsys', 'package', packageName], { serial });
    return parseAppDetail(packageName, result.stdout);
  }

  async install(serial: string, apkPath: string): Promise<string> {
    const result = await this.run(['install', '-r', '-g', apkPath], { serial, timeoutMs: 120_000 });
    const combined = combinedOutput(result);
    if (result.exitCode !== 0 || combined.includes('Failure')) {
      throw new AdbCommandError(combined);
    }
    return combined;
  }

  async uninstall(serial: string, packageName: string): Promise<void> {
    const result = await this.run(['uninstall', packageName], { serial });
    const combined = combinedOutput(result);
    if (result.exitCode !== 0 || combined.includes('Failure')) {
      throw new AdbCommandError(combined);
    }
  }

  async forceStop(serial: string, packageName: string): Promise<void> {
    await this.run(['shell', 'am', 'force-stop', packageName], { serial });
  }

  async clearData(serial: string, packageName: string): Promise<void> {
    const result = await this.run(['shell', 'pm', 'clear', packageName], { serial });
    if (result.stdout.includes('Failed')) {
      throw new AdbCommandError(combinedOutput(result));
    }
  }

  async setEnabled(serial: string, packageName: string, enabled: boolean): Promise<void> {
    const subcommand = enabled ? 'enable' : 'disable-user';
    const args = ['shell', 'pm', subcommand];
    if (!enabled) args.push('--user', '0');
    args.push(packageName);
    const result = await this.run(args, { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  async grantPermission(serial: string, packageName: string, permission: string): Promise<void> {
    const result = await this.run(['shell', 'pm', 'grant', packageName, permission], { serial });
    if (result.exitCode !== 0 || result.stderr.length > 0) {
      throw new AdbCommandError(combinedOutput(result));
    }
  }

  async revokePermission(serial: string, packageName: string, permission: string): Promise<void> {
    const result = await this.run(['shell', 'pm', 'revoke', packageName, permission], { serial });
    if (result.exitCode !== 0 || result.stderr.length > 0) {
      throw new AdbCommandError(combinedOutput(result));
    }
  }

  // MARK: Файлы устройства

  async listDirectory(serial: string, dirPath: string): Promise<RemoteFile[]> {
    const result = await this.run(['shell', 'ls', '-la', dirPath], { serial });
    if (result.stdout.length === 0 && result.stderr.length > 0) {
      throw new AdbCommandError(result.stderr);
    }
    return parseRemoteFiles(result.stdout, dirPath);
  }

  async push(serial: string, localPath: string, remotePath: string): Promise<void> {
    const result = await this.run(['push', localPath, remotePath], { serial, timeoutMs: 300_000 });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  async pull(serial: string, remotePath: string, localPath: string): Promise<void> {
    const result = await this.run(['pull', remotePath, localPath], { serial, timeoutMs: 300_000 });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  async makeDirectory(serial: string, dirPath: string): Promise<void> {
    const result = await this.run(['shell', 'mkdir', '-p', dirPath], { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  async removeRemote(serial: string, targetPath: string, recursive: boolean): Promise<void> {
    const args = ['shell', 'rm', recursive ? '-rf' : '-f', targetPath];
    const result = await this.run(args, { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
  }

  // MARK: Мониторинг

  async deviceStats(serial: string): Promise<DeviceStats> {
    const [cpu, mem, battery] = await Promise.all([
      this.run(['shell', 'dumpsys', 'cpuinfo'], { serial }),
      this.run(['shell', 'cat', '/proc/meminfo'], { serial }),
      this.run(['shell', 'dumpsys', 'battery'], { serial }),
    ]);
    return parseDeviceStats(combinedOutput(cpu), combinedOutput(mem), combinedOutput(battery));
  }

  async runningProcesses(serial: string): Promise<RunningProcess[]> {
    const result = await this.run(['shell', 'ps', '-A', '-o', 'PID,PPID,USER,RSS,NAME'], { serial });
    return parseProcessList(result.stdout);
  }

  async killProcess(serial: string, pid: number): Promise<void> {
    await this.run(['shell', 'kill', String(pid)], { serial });
  }

  // MARK: Безопасность устройства

  /** Аналог ADBService.securityInfo(serial:) — локальные признаки
   * целостности устройства (root/разлочка/debuggable). Полноценный
   * SafetyNet/Play Integrity с устройства через adb не выполнить, это
   * удалённая проверка на серверах Google. */
  async securityInfo(serial: string): Promise<DeviceSecurityInfo> {
    const [verifiedBoot, flashLocked, debuggable, secure, suCheck, playProtect] = await Promise.all([
      this.run(['shell', 'getprop', 'ro.boot.verifiedbootstate'], { serial }),
      this.run(['shell', 'getprop', 'ro.boot.flash.locked'], { serial }),
      this.run(['shell', 'getprop', 'ro.debuggable'], { serial }),
      this.run(['shell', 'getprop', 'ro.secure'], { serial }),
      this.run(['shell', 'which', 'su'], { serial }),
      this.run(['shell', 'settings', 'get', 'global', 'package_verifier_user_consent'], { serial }),
    ]);
    const trim = (r: ProcessResult) => r.stdout.trim();

    const verifiedBootValue = trim(verifiedBoot);
    const flashLockedValue = trim(flashLocked);
    const debuggableValue = trim(debuggable);
    const secureValue = trim(secure);
    const suValue = trim(suCheck);
    const playProtectValue = trim(playProtect);

    return {
      verifiedBootState: verifiedBootValue.length === 0 ? undefined : verifiedBootValue,
      bootloaderLocked: flashLockedValue.length === 0 ? undefined : flashLockedValue === '1',
      isDebuggable: debuggableValue === '1',
      isSecure: secureValue !== '0',
      suBinaryPresent: suValue.length > 0 && !suValue.toLowerCase().includes('not found'),
      playProtectConsent:
        playProtectValue.length === 0 || playProtectValue.toLowerCase() === 'null' ? undefined : playProtectValue,
    };
  }

  // MARK: Сетевой трафик по приложению

  async networkUsage(serial: string, uid: number): Promise<NetworkUsage> {
    const result = await this.run(['shell', 'dumpsys', 'netstats', 'detail'], { serial });
    return parseNetworkUsage(combinedOutput(result), uid);
  }

  // MARK: Экранное время приложений

  async usageStats(serial: string): Promise<AppUsageStat[]> {
    const result = await this.run(['shell', 'dumpsys', 'usagestats'], { serial });
    return parseUsageStats(combinedOutput(result));
  }

  // MARK: ANR / tombstones

  /** Список файлов в /data/anr/ и /data/tombstones/. Без root оба каталога
   * обычно недоступны (Permission denied) — в этом случае просто пропускаем
   * соответствующую директорию, а не считаем это ошибкой всей операции. */
  async crashTraces(serial: string): Promise<CrashTraceFile[]> {
    const [anrResult, tombResult] = await Promise.all([
      this.run(['shell', 'ls', '-1', '/data/anr/'], { serial }).catch(() => undefined),
      this.run(['shell', 'ls', '-1', '/data/tombstones/'], { serial }).catch(() => undefined),
    ]);
    const files: CrashTraceFile[] = [];
    if (anrResult && anrResult.exitCode === 0) {
      files.push(...parseCrashTraceListing(anrResult.stdout, '/data/anr/', 'anr'));
    }
    if (tombResult && tombResult.exitCode === 0) {
      files.push(...parseCrashTraceListing(tombResult.stdout, '/data/tombstones/', 'tombstone'));
    }
    return files;
  }

  /** Хвост файла трейса — полные tombstone-файлы могут быть большими,
   * показываем последние ~30000 байт, обычно там самое важное (стек, сигнал). */
  async readCrashTrace(serial: string, filePath: string): Promise<string> {
    const result = await this.run(['shell', 'tail', '-c', '30000', filePath], { serial });
    if (result.exitCode !== 0) throw new AdbCommandError(combinedOutput(result));
    return result.stdout;
  }
}
